### PR TITLE
[Contract]Add WooStakingVault

### PR DESCRIPTION
**Pull request by merlincdj**
- [x] withdraw function

### CODE REVIEW

**Comment by merlincdj** [comment]
Using hardhat deploy and verify on bsc testnet
https://testnet.bscscan.com/address/0xAaf8F6F1F64B29fAdF5320268A45074E2154dec7#code

File: contracts/WooStakingVault.sol [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+
+    IERC20 public token;
+
+    address public governance;
+    address public controller;
+
+    constructor(address _token)
+        public
+        ERC20(
+            string(abi.encodePacked("x", ERC20(_token).name())),
```
Comment (skyinthesea-woo) [inline contracts/WooStakingVault.sol]: "x" --->  "Interest bearing"

File: contracts/WooStakingVault.sol [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+        token = IERC20(_token);
+        governance = owner();
+    }
+
+    /* ----- Write Functions ----- */
+
+    function deposit(uint256 _amount) external {
+        uint256 _pool = balance();
+        uint256 _before = token.balanceOf(address(this));
+        token.safeTransferFrom(msg.sender, address(this), _amount);
```
Comment (skyinthesea-woo) [inline contracts/WooStakingVault.sol:70]: Please utilize our new TransferHelper method:

There's sample code in WooRouter.sol

Comment (skyinthesea-woo) [inline contracts/WooStakingVault.sol:70]: Please utilize our new TransferHelper method:

There's sample code in WooRouter.sol

File: contracts/WooStakingVault.sol [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+        )
+    {
+        token = IERC20(_token);
+        governance = owner();
+    }
+
+    /* ----- Write Functions ----- */
+
+    function deposit(uint256 _amount) external {
+        uint256 _pool = balance();
```
Comment (skyinthesea-woo) [inline contracts/WooStakingVault.sol:68]: poolBalance

File: contracts/WooStakingVault.sol [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+        uint256 _before = token.balanceOf(address(this));
```
Comment (skyinthesea-woo) [inline contracts/WooStakingVault.sol:69]: balanceBefore

File: contracts/WooStakingVault.sol [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+*/
+
+import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import "@openzeppelin/contracts/token/ERC20/ERC20.sol";
+import "@openzeppelin/contracts/token/ERC20/SafeERC20.sol";
+import "@openzeppelin/contracts/math/SafeMath.sol";
+import "@openzeppelin/contracts/utils/Address.sol";
+import '@openzeppelin/contracts/access/Ownable.sol';
+
+contract wVault is ERC20, Ownable {
```
Comment (skyinthesea-woo) [inline contracts/WooStakingVault.sol:44]: WooStakingVault

Contract name's first character needs to be "UpperCase"

File: contracts/WooStakingVault.sol [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+    using SafeERC20 for IERC20;
+    using Address for address;
+    using SafeMath for uint256;
+
+    IERC20 public token;
```
Comment (skyinthesea-woo) [inline contracts/WooStakingVault.sol:49]: stakedToken

File: contracts/WooStakingVault.sol [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+
+    address public governance;
+    address public controller;
```
Comment (skyinthesea-woo) [inline contracts/WooStakingVault.sol:52]: Not needed anymore

File: contracts/WooStakingVault.sol [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+        uint256 _before = token.balanceOf(address(this));
+        token.safeTransferFrom(msg.sender, address(this), _amount);
+        uint256 _after = token.balanceOf(address(this));
+        _amount = _after.sub(_before);
+        uint256 shares = 0;
+        if (totalSupply() == 0) {
+            shares = _amount;
+        } else {
+            shares = (_amount.mul(totalSupply())).div(_pool);
+        }
```
Comment (skyinthesea-woo) [inline contracts/WooStakingVault.sol:78]: uint256 totalSupply = totalSupply();
uint256 shares = totalSupply == 0 ? _amount : _amount.mul(totalSupply).div(poolBalance)

File: contracts/WooStakingVault.sol [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+        _mint(msg.sender, shares);
+    }
+
+    function withdraw(uint256 _shares) external {
+        uint256 r = (balance().mul(_shares)).div(totalSupply());
```
Comment (skyinthesea-woo) [inline contracts/WooStakingVault.sol:83]: r -> shareBalance

或者换一个你觉得有意义的名字； 别用  r

**Review by skyinthesea-woo — CHANGES_REQUESTED** [review]
A bunch of comments to fix.

**Review by skyinthesea-woo — COMMENTED** [review]
Looks good.   

再加上unit tests就完美了

File: contracts/WooStakingVault.sol [hunk 10]
```diff
@@ -34,76 +34,105 @@ pragma solidity 0.6.12;
...
         ERC20(
-            string(abi.encodePacked('x', ERC20(_token).name())),
-            string(abi.encodePacked('x', ERC20(_token).symbol()))
+            string(abi.encodePacked("Interest bearing", ERC20(_stakedToken).name())),
+            string(abi.encodePacked("x", ERC20(_stakedToken).symbol()))
         )
     {
-        token = IERC20(_token);
-        governance = owner();
+        stakedToken = IERC20(_stakedToken);
```
Comment (skyinthesea-woo) [inline contracts/WooStakingVault.sol:80]: address 的参数 require != address(0) 

可以看 WooRouter 的代码。

File: contracts/WooStakingVault.sol [hunk 11]
```diff
@@ -34,76 +34,105 @@ pragma solidity 0.6.12;
...
+    /* ----- External Functions ----- */
+
+    function deposit(uint256 _amount) external onlyHuman {
+        uint256 poolBalance = balance();
+        uint256 balanceBefore = stakedToken.balanceOf(address(this));
+        TransferHelper.safeTransferFrom(address(stakedToken), msg.sender, address(this), _amount);
+        uint256 balanceAfter = stakedToken.balanceOf(address(this));
+        _amount = balanceAfter.sub(balanceBefore);
+
+        uint256 xTotalSupply = totalSupply();
```
Comment (skyinthesea-woo) [inline contracts/WooStakingVault.sol:92]: nice naming for this local variable.

File: contracts/WooStakingVault.sol [hunk 12]
```diff
@@ -34,76 +34,105 @@ pragma solidity 0.6.12;
...
+    }
+
+    /* ----- Private Functions ----- */
+
+    function _updateCostSharePrice(uint256 _amount, uint256 _shares) private {
+        uint256 beforeShares = balanceOf(msg.sender);
+        uint beforeCost = costSharePrice[msg.sender];
+        uint afterCost = (
+            beforeShares.mul(beforeCost).add(_amount.mul(1e18))
+        ).div(beforeShares.add(_shares));
```
Comment (skyinthesea-woo) [inline contracts/WooStakingVault.sol]: 能不能这里写一下comment ； 这个公式是什么？

这样方便后面进行维护。

File: contracts/WooStakingVault.sol [hunk 13]
```diff
@@ -34,103 +34,159 @@ pragma solidity 0.6.12;
...
+        uint256 lastReserveWithdrawTime; // keeps track of reverseWithdraw time for potential penalty
+    }
+
     /* ----- State variables ----- */
 
     IERC20 public stakedToken;
-    IWooCoolDownVault public wooCoolDownVault;
-    mapping(address => uint256) public costSharePrice;
+    mapping (address => uint256) public costSharePrice;
+    mapping(address => UserInfo) public userInfo;
```
Comment (skyinthesea-woo) [inline contracts/WooStakingVault.sol:60]: Pay attention to the code format.

File: contracts/WooStakingVault.sol [hunk 14]
```diff
@@ -34,103 +34,159 @@ pragma solidity 0.6.12;
...
+
+    uint256 public totalReserveAmount = 0; // affected by reserveWithdraw and withdraw
+    uint256 public withdrawFeePeriod = 72 hours; // 3 days
+    uint256 public withdrawFee = 10; // 0.1%
```
Comment (skyinthesea-woo) [inline contracts/WooStakingVault.sol:64]: 稍微注释下单位

File: contracts/WooStakingVault.sol [hunk 15]
```diff
@@ -34,103 +34,159 @@ pragma solidity 0.6.12;
...
 
+        // must be execute before _mint
         _updateCostSharePrice(_amount, shares);
 
         _mint(msg.sender, shares);
     }
 
-    function withdraw(uint256 _shares) external onlyHuman {
-        uint256 withdrawBalance = (balance().mul(_shares)).div(totalSupply());
+    function reserveWithdraw(uint256 _shares) external whenNotPaused {
```
Comment (skyinthesea-woo) [inline contracts/WooStakingVault.sol:101]: 我建议别用 _shares 吧

就用  xWooAmount  ； _shares 都是之前 YFI 的遗留产物，而且很疑惑人。

File: contracts/WooStakingVault.sol [hunk 16]
```diff
@@ -34,103 +34,159 @@ pragma solidity 0.6.12;
...
     /* ----- External Functions ----- */
 
-    function deposit(uint256 _amount) external onlyHuman {
-        uint256 poolBalance = balance();
-        uint256 balanceBefore = stakedToken.balanceOf(address(this));
+    function deposit(uint256 _amount) external whenNotPaused {
+        uint256 balanceBefore = balance();
         TransferHelper.safeTransferFrom(address(stakedToken), msg.sender, address(this), _amount);
-        uint256 balanceAfter = stakedToken.balanceOf(address(this));
+        uint256 balanceAfter = stakedToken.balanceOf(address(this)).sub(totalReserveAmount);
```
Comment (skyinthesea-woo) [inline contracts/WooStakingVault.sol:89]: call "balance()" please.

File: test/WooStakingVault.test.ts [hunk 17]
```diff
@@ -138,7 +151,154 @@ describe('WooStakingVault', () => {
...
+  let wooToken: TestToken
+
+  let onlyOwnerRevertedMessage: string
+  let setWithdrawFeePeriodExceedMessage: string
+  let setWithdrawFeeExceedMessage: string
+  let whenNotPausedRevertedMessage: string
+  let whenPausedRevertedMessage: string
+
+  before(async () => {
+    ;[owner, user, treasury, newTreasury] = await ethers.getSigners()
```
Comment (skyinthesea-woo) [inline test/WooStakingVault.test.ts:211]: 这里的 ; 是什么 ？

**Comment by skyinthesea-woo** [comment]
这边跑了之后 coverage 是 100% 吗？  （Vault这个class）

**Comment by skyinthesea-woo** [comment]
Have a few inline comments. Others look good. 

We can do a smart contract security check on Monday.

File: contracts/WooStakingVault.sol (hunk 13 above)
Comment (merlincdj) [reply contracts/WooStakingVault.sol:60]: resolved by lint-action

File: test/WooStakingVault.test.ts (hunk 17 above)
Comment (merlincdj) [reply test/WooStakingVault.test.ts:211]: lint-action make this. # 9e65944b2ad339d73060fa61e335631978a14b98

File: contracts/WooStakingVault.sol (hunk 15 above)
Comment (merlincdj) [reply contracts/WooStakingVault.sol:101]: shares涉及到share price(单位净值)和cost share price(用户成本单位净值)，我建议不改动

**Review by skyinthesea-woo — APPROVED** [review]

**Comment by merlincdj** [comment]
> 这边跑了之后 coverage 是 100% 吗？ （Vault这个class）

only line 115 not coverage, maybe it's impossible situation, keep this logic for safety.

**Pull request merged**